extensions/avro-extensions/src/main/java/io/druid/data/input/AvroStreamInputRowParser.java

```diff
@@ -0,0 +1,116 @@
+import io.druid.data.input.impl.TimestampSpec;
+import io.druid.data.input.schemarepo.SubjectAndIdConverter;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.util.ByteBufferInputStream;
+import org.joda.time.DateTime;
+import org.schemarepo.Repository;
+import org.schemarepo.api.TypedSchemaRepository;
+import org.schemarepo.api.converter.AvroSchemaConverter;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+
+public class AvroStreamInputRowParser<SUBJECT, ID> implements ByteBufferInputRowParser
+{
+  private final ParseSpec parseSpec;
+  private final List<String> dimensions;
+
+  private final TypedSchemaRepository<ID, Schema, SUBJECT> typedRepository;
+  private final SubjectAndIdConverter<SUBJECT, ID> subjectAndIdConverter;
+  private final Repository schemaRepository;
+
+  @JsonCreator
+  public AvroStreamInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("subjectAndIdConverter") SubjectAndIdConverter<SUBJECT, ID> subjectAndIdConverter,
+      @JsonProperty("schemaRepository") Repository schemaRepository
+  )
+  {
+    this.parseSpec = parseSpec;
+    this.dimensions = parseSpec.getDimensionsSpec().getDimensions();
+    this.subjectAndIdConverter = subjectAndIdConverter;
+    this.schemaRepository = schemaRepository;
+    typedRepository = new TypedSchemaRepository<ID, Schema, SUBJECT>(
+        schemaRepository,
+        subjectAndIdConverter.getIdConverter(),
+        new AvroSchemaConverter(false),
+        subjectAndIdConverter.getSubjectConverter()
+    );
+  }
+
+  @Override
+  public InputRow parse(ByteBuffer input)
+  {
+    ByteBufferInputStream inputStream = new ByteBufferInputStream(Collections.singletonList(input));
+    Pair<SUBJECT, ID> subjectAndId = subjectAndIdConverter.toSubjectAndId(input);
+    Schema schema = typedRepository.getSchema(subjectAndId.lhs, subjectAndId.rhs);
+    DatumReader<GenericRecord> reader = new GenericDatumReader<GenericRecord>(schema);
+    try {
+      GenericRecord record = reader.read(null, DecoderFactory.get().binaryDecoder(inputStream, null));
+      return doParse(record, parseSpec, dimensions);
+    }
+    catch (IOException e) {
+      throw new RuntimeException("Fail to decode avro message!", e);
+    }
+  }
+
+  protected static InputRow doParse(
+      GenericRecord record, ParseSpec parseSpec, List<String> dimensions
+  )
+  {
+    TimestampSpec timestampSpec = parseSpec.getTimestampSpec();
+    DateTime dateTime = timestampSpec.extractTimestamp(
+        Collections.singletonMap(
+            timestampSpec.getTimestampColumn(),
+            record.get(timestampSpec.getTimestampColumn())
+        )
+    );
+    return new MapBasedInputRow(dateTime, dimensions, new GenericRecordAsMap(record));
```

Comment:
GenericRecordAsMap is fine for primitives, but we should handle ARRAY for multivalued dimensions, UNION in case schema was something like ["null","int"] instead of just "int" (pig AvroStorage would generate schema like that).
you could write a map converter something like below (we've been using something similar in our production for a while)
```
  protected static Map<String, Object> genericRecordToMap(GenericRecord genericData) throws IOException
  {
    Map<String, Object> data = new HashMap<String, Object>();
    List<Schema.Field> fields = genericData.getSchema().getFields();
    for(Schema.Field field : fields) {
      String name = field.name();
      Object value = genericData.get(name);
      data.put(name, value);
      if(value != null) {
        Schema.Type type = field.schema().getType();
        if(type == Schema.Type.UNION) { //handling primitive types defined in schema like ["null","int"]
          Schema fieldSchema = field.schema();
          int index = GenericData.get().resolveUnion(fieldSchema, value);
          type = fieldSchema.getTypes().get(index).getType();
        }
        if(type ==  Schema.Type.STRING) {
          data.put(name, value.toString());
        }
        if(type ==  Schema.Type.ARRAY) { //handling multivalued dimensions
          List<String> dimValues = new ArrayList<String>();

          for(Object dimValue : (Collection)value) {
            Object subvalue = dimValue;

            if(dimValue instanceof GenericRecord) {
              /* This addresses the way pig AvroStorage outputs an array/
              * bag of values. */
              subvalue = subvalue = ((GenericRecord)dimValue).get(name);
              if (subvalue == null) {
                throw new IOException("Could not process bag " + name + ", it's elements should be named " + name);
              }
            }

            dimValues.add(subvalue.toString());
          }
          data.put(name, dimValues);
        }
      }
    }
    return data;
  }

```